File: src/main/java/app/voltlauncher/voltlauncher/ProtocolHandler.java
package app.voltlauncher.voltlauncher;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.function.Consumer;

public final class ProtocolHandler {

    private static ServerSocket serverSocket;
    private static Consumer < String> callbackHandler;

    private ProtocolHandler() {}

    public static void startListening(int port, Consumer < String> onCallback) {
        callbackHandler = onCallback;

        Thread thread = new Thread(() -> {
            try {
                serverSocket = new ServerSocket(port);
                System.out.println("Protocol handler listening on port " + port);

                while (!Thread.currentThread().isInterrupted()) {
                    Socket client = serverSocket.accept();
                    if (client == null) {
                        break;
                    }

                    Thread clientThread = new Thread(() -> {
                        try (BufferedReader reader = new BufferedReader( new InputStreamReader(client.getInputStream()));
                        OutputStream out = client.getOutputStream()) {

                            String requestLine = reader.readLine();
                            if (requestLine != null && requestLine.startsWith("GET ")) {
                                String[] parts = requestLine.split(" ");
                                String path = parts.length > 1 ? parts[1] : "/";

                                String response =
                                "HTTP/1.1 200 OK\r\n" +
                                "Content-Type: text/html\r\n" +
                                "\r\n" +
                                "You can close this window and return to the application.";

                                out.write(response.getBytes());
                                out.flush();

                                if (callbackHandler != null) {
                                    callbackHandler.accept(path);
                                }
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                        } finally {
                            try {
                                client.close();
                            } catch (IOException ignored) {
                            }
                        }
                    });
                    clientThread.setDaemon(true);
                    clientThread.start();
                }
            } catch (Exception e) {
                if (!Thread.currentThread().isInterrupted()) {
                    e.printStackTrace();
                }
            }
        });

        thread.setDaemon(true);
        thread.start();
    }

    public static void stop() {
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (IOException ignored) {
        } finally {
            serverSocket = null;
        }
    }
}

File: src/main/kotlin/de/eztxm/thelauncherproject/ProtocolHandler.kt
package de.eztxm.thelauncherproject

import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.ServerSocket

object ProtocolHandler {
    private var serverSocket: ServerSocket? = null
    private var callbackHandler: ((String) -> Unit)? = null
    
    fun startListening(port: Int = 7071, onCallback: (String) -> Unit) {
        callbackHandler = onCallback
        
        Thread {
            try {
                serverSocket = ServerSocket(port)
                println("Protocol handler listening on port $port")
                
                while (!Thread.currentThread().isInterrupted) {
                    val client = serverSocket?.accept() ?: break
                    
                    Thread {
                        try {
                            val reader = BufferedReader(InputStreamReader(client.getInputStream()))
                            val requestLine = reader.readLine()
                            
                            if (requestLine != null && requestLine.startsWith("GET ")) {
                                val path = requestLine.split(" ")[1]
                                
                                // Send HTTP response
                                val response = """
                                    HTTP/1.1 200 OK
                                    Content-Type: text/html
                                    
                                    <html>
                                    <body>
                                        <h1>Authentication Successful!</h1>
                                        <p>You can close this window and return to the application.</p>
                                        <script>window.close();</script>
                                    </body>
                                    </html>
                                """.trimIndent()
                                
                                client.getOutputStream().write(response.toByteArray())
                                client.getOutputStream().flush()
                                
                                // Notify callback
                                callbackHandler?.invoke(path)
                            }
                            
                            client.close()
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }.start()
                }
            } catch (e: Exception) {
                if (!Thread.currentThread().isInterrupted) {
                    e.printStackTrace()
                }
            }
        }.apply {
            isDaemon = true
            start()
        }
    }
    
    fun stop() {
        serverSocket?.close()
        serverSocket = null
    }
}
